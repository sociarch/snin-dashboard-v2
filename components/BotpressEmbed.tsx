"use client";

import { useEffect } from "react";

export default function BotpressEmbed() {
    useEffect(() => {
        // Load the Botpress script
        const script = document.createElement("script");
        script.src = "https://cdn.botpress.cloud/webchat/v2.2/inject.js";
        script.async = true;

        // Add event listener for script load
        script.onload = () => {
            // Load the custom script
            const customScript = document.createElement("script");
            customScript.src = "https://files.bpcontent.cloud/2024/10/21/06/20241021062910-6QRSUQCY.js";
            customScript.async = true;

            // Add event listeners after custom script is loaded
            // customScript.onload = addEventListeners;

            document.body.appendChild(customScript);
        };

        document.body.appendChild(script);

        return () => {
            // Clean up the scripts when the component unmounts
            const scripts = document.querySelectorAll('script[src^="https://cdn.botpress.cloud"], script[src^="https://files.bpcontent.cloud"]');
            scripts.forEach((script) => script.remove());
        };
    }, []);

    return null; // This component doesn't render anything visible
}
