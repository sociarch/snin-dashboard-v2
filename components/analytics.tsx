"use client";

export function Analytics() {
  return (
    <>
      {/* Google Analytics */}
      <script
        async
        src="https://www.googletagmanager.com/gtag/js?id=G-3TXCQD4CD9"
      />
      <script
        dangerouslySetInnerHTML={{
          __html: `
            window.dataLayer = window.dataLayer || [];
            function gtag(){dataLayer.push(arguments);}
            gtag('js', new Date());
            gtag('config', 'G-3TXCQD4CD9');
          `,
        }}
      />

      {/* Microsoft Clarity */}
      <script
        dangerouslySetInnerHTML={{
          __html: `
            (function(c,l,a,r,i,t,y){
              c[a]=c[a]||function(){(c[a].q=c[a].q||[]).push(arguments)};
              t=l.createElement(r);t.async=1;t.src="https://www.clarity.ms/tag/"+i;
              y=l.getElementsByTagName(r)[0];y.parentNode.insertBefore(t,y);
            })(window,document,"clarity","script","nkezm8mdse");
          `,
        }}
      />
    </>
  );
} 