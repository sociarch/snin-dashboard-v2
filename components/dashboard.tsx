"use client";

import { useState, useEffect, useRef } from "react";
import { Button } from "@/components/ui/button";
import { Table, TableBody, TableCell, TableHead, TableHeader, TableRow } from "@/components/ui/table";
import { Card, CardContent, CardDescription, CardHeader, CardTitle } from "@/components/ui/card";
import { Input } from "@/components/ui/input";
import * as d3 from "d3";
import { pollData, PollDataItem } from "./pollData";
import { ThemeToggle } from "./theme-toggle";

export function Dashboard() {
    // State variables
    const [selectedPoll, setSelectedPoll] = useState<PollDataItem | null>(null);
    const [searchTerm, setSearchTerm] = useState("");
    const [filteredData, setFilteredData] = useState<PollDataItem[]>(pollData);
    const chartContainerRef = useRef<HTMLDivElement>(null);
    const [chartDimensions, setChartDimensions] = useState({ width: 0, height: 0 });

    // Add new state to control visibility
    const [showTitle, setShowTitle] = useState(false);
    const [showPercentages, setShowPercentages] = useState(false);
    const [showAnnotation, setShowAnnotation] = useState(false);

    // Add new state to handle loading
    const [isLoading, setIsLoading] = useState(false);
    const timeoutRef = useRef<NodeJS.Timeout | null>(null);

    // Handler for when a table row is clicked
    const handleRowClick = (poll: PollDataItem) => {
        if (poll !== selectedPoll) {
            setIsLoading(true);

            if (timeoutRef.current) {
                clearTimeout(timeoutRef.current);
            }

            // Increase the delay before updating the poll
            timeoutRef.current = setTimeout(() => {
                setSelectedPoll(poll);
                // Keep isLoading true to allow for chart creation
            }, 500); // Increased from 300ms to 500ms
        }
    };

    // Handler for search input changes
    const handleSearch = (event: React.ChangeEvent<HTMLInputElement>) => {
        const searchTerm = event.target.value.toLowerCase();
        setSearchTerm(searchTerm);
        // Filter the poll data based on the search term
        const filtered = pollData.filter(
            (poll) =>
                poll.caption.toLowerCase().includes(searchTerm) ||
                poll.option1.toLowerCase().includes(searchTerm) ||
                poll.option2.toLowerCase().includes(searchTerm) ||
                poll.resp_option1.toString().includes(searchTerm) ||
                poll.resp_option2.toString().includes(searchTerm)
        );
        setFilteredData(filtered);
    };

    // Function to update chart dimensions
    const updateChartDimensions = () => {
        if (chartContainerRef.current) {
            const width = chartContainerRef.current.clientWidth;
            const height = Math.min(width / 2, 500); // Maintain proportion, but limit the height
            setChartDimensions({ width, height });
        }
    };

    // Effect to handle window resize and initial dimension calculation
    useEffect(() => {
        updateChartDimensions();
        window.addEventListener("resize", updateChartDimensions);
        return () => window.removeEventListener("resize", updateChartDimensions);
    }, []);

    // Function to create the D3 chart
    const createChart = (poll: PollDataItem) => {
        if (!chartContainerRef.current) return;

        const containerWidth = chartContainerRef.current.clientWidth;
        const svgWidth = containerWidth * 0.95; // 95% of the container width
        const svgHeight = svgWidth / 2; // Half the width for a semi-circle

        // Clear existing chart
        d3.select(chartContainerRef.current).selectAll("*").remove();

        const svg = d3
            .select(chartContainerRef.current)
            .append("svg")
            .attr("width", svgWidth)
            .attr("height", svgHeight)
            .style("display", "block")
            .style("margin", "auto") // Center the SVG horizontally
            .append("g")
            .attr("transform", `translate(${svgWidth / 2}, ${svgHeight})`);

        // Define arc generators
        const arc = d3
            .arc<d3.PieArcDatum<any>>()
            .innerRadius(svgHeight * 0.6)
            .outerRadius(svgHeight);

        const labelArc = d3
            .arc<d3.PieArcDatum<any>>()
            .innerRadius(svgHeight * 0.85)
            .outerRadius(svgHeight * 0.85);

        // Define pie layout
        const pie = d3
            .pie<any>()
            .sort(null)
            .value((d) => d.value)
            .startAngle(-Math.PI / 2)
            .endAngle(Math.PI / 2);

        // Prepare data for the pie chart
        const data = [
            { label: poll.option1, value: parseInt(poll.resp_option1), color: "#ffd700" },
            { label: poll.option2, value: parseInt(poll.resp_option2), color: "#FA8072" },
        ];

        const total = data.reduce((acc, d) => acc + d.value, 0);

        // Create arc groups and paths
        const arcs = svg.selectAll(".arc").data(pie(data)).enter().append("g").attr("class", "arc");

        arcs.append("path")
            .attr("d", arc)
            .style("fill", (d) => d.data.color)
            .transition()
            .duration(1000)
            .attrTween("d", function (d) {
                const interpolate = d3.interpolate({ startAngle: -Math.PI / 2, endAngle: -Math.PI / 2 }, d);
                return function (t) {
                    return arc(interpolate(t));
                };
            });

        // Add label paths
        arcs.append("path")
            .attr("class", "label-path")
            .attr("id", (d, i) => `label-path-${i}`)
            .attr("d", labelArc)
            .style("fill", "none")
            .style("stroke", "none");

        // Add labels
        arcs.append("text")
            .append("textPath")
            .attr("xlink:href", (d, i) => `#label-path-${i}`)
            .attr("startOffset", (d, i) => (i === 0 ? "3%" : "47%"))
            .text((d) => `${d.data.label}`)
            .style("text-anchor", (d, i) => (i === 0 ? "start" : "end"))
            .style("font-family", "'Roboto', sans-serif")
            .style("font-size", "1.1em") // Increased from 0.875em to 1.1em
            .style("fill", "black") // Changed from white to black
            .attr("dy", "0.35em")
            .style("opacity", 0)
            .transition()
            .delay(1000)
            .duration(500)
            .style("opacity", 1);
    };

    // Function to calculate p-value (simplified approximation)
    const calculatePValue = (data: { value: number }[]) => {
        const n = data.reduce((acc, d) => acc + d.value, 0);
        const p1 = data[0].value / n;
        const p2 = data[1].value / n;
        const p = (data[0].value + data[1].value) / (2 * n);

        const z = (p1 - p2) / Math.sqrt(p * (1 - p) * (1 / n + 1 / n));
        const pValue = Math.exp(-0.717 * z - 0.416 * z * z);

        return pValue;
    };

    // Effect to create/update chart when selectedPoll or chartDimensions change
    useEffect(() => {
        if (selectedPoll && chartDimensions.width > 0 && isLoading) {
            createChart(selectedPoll);

            // Reset visibility states
            setShowTitle(false);
            setShowPercentages(false);
            setShowAnnotation(false);

            // Delay the start of fade-in effects and set isLoading to false
            setTimeout(() => {
                setIsLoading(false);
                setTimeout(() => setShowTitle(true), 300);
                setTimeout(() => setShowPercentages(true), 600);
                setTimeout(() => setShowAnnotation(true), 900);
            }, 500); // Allow 500ms for chart creation before starting fade-ins
        }
    }, [selectedPoll, chartDimensions, isLoading]);

    // Clean up the timeout on component unmount
    useEffect(() => {
        return () => {
            if (timeoutRef.current) {
                clearTimeout(timeoutRef.current);
            }
        };
    }, []);

    // Handler for button clicks (placeholder for future functionality)
    const handleButtonClick = (buttonId: string) => {
        console.log(`Button ${buttonId} clicked`);
        // Placeholder for future functionality
    };

    return (
        <div id="dashboard-container" className="flex flex-col h-screen">
            <header id="dashboard-header" className="flex-none p-4 flex justify-between items-center">
                <h1 className="text-2xl font-bold">Snapinput</h1>
                <ThemeToggle />
            </header>
            <main id="dashboard-main" className="flex-grow overflow-hidden p-4">
                <div id="dashboard-grid" className="grid grid-cols-1 lg:grid-cols-[minmax(auto,600px)_1fr] gap-4 h-full">
                    <Card id="chart-card" className="flex flex-col bg-black text-white w-full max-w-[600px] mx-auto">
                        <CardContent
                            id="chart-content"
                            className={`flex-grow flex flex-col h-full max-h-[800px] transition-opacity duration-500 ${
                                isLoading ? "opacity-0" : "opacity-100"
                            }`}
                        >
                            <div
                                id="chart-title"
                                className={`flex-1 flex items-center justify-center transition-opacity duration-500 ${showTitle ? "opacity-100" : "opacity-0"}`}
                            >
                                <h2 className="text-3xl font-bold text-center px-4">
                                    {selectedPoll ? selectedPoll.caption : "Click a row to see detailed results"}
                                </h2>
                            </div>
                            <div id="chart-container" ref={chartContainerRef} className="flex-1 flex justify-center items-center">
                                {/* SVG will be inserted here */}
                            </div>
                            <div
                                id="chart-percentages"
                                className={`flex justify-center items-center transition-opacity duration-500 ${showPercentages ? "opacity-100" : "opacity-0"}`}
                            >
                                <table className="w-[95%] mx-auto">
                                    <tbody>
                                        <tr className="w-full">
                                            <td className="w-1/2 text-left font-['Roboto'] text-[2em] font-bold" style={{ color: "#ffd700" }}>
                                                {selectedPoll ? `${(parseFloat(selectedPoll.pct_option1) * 100).toFixed(0)}%` : ""}
                                            </td>
                                            <td className="w-1/2 text-right font-['Roboto'] text-[2em] font-bold" style={{ color: "#FA8072" }}>
                                                {selectedPoll ? `${(parseFloat(selectedPoll.pct_option2) * 100).toFixed(0)}%` : ""}
                                            </td>
                                        </tr>
                                    </tbody>
                                </table>
                            </div>
                            <div
                                id="chart-annotation"
                                className={`flex justify-center items-center py-4 transition-opacity duration-500 ${
                                    showAnnotation ? "opacity-100" : "opacity-0"
                                }`}
                            >
                                <span id="chart-footer-center" className="text-[1.5em] text-center">
                                    Most people prefer{" "}
                                    {selectedPoll &&
                                        (parseInt(selectedPoll.resp_option1) > parseInt(selectedPoll.resp_option2)
                                            ? selectedPoll.option1
                                            : selectedPoll.option2)}
                                </span>
                            </div>
                            <div id="chart-footer" className="flex-1 flex justify-between items-end px-2">
                                <span id="chart-footer-left" className="text-xs">
                                    snapinput.com
                                </span>
                                <span id="chart-footer-right" className="text-xs">
                                    {selectedPoll &&
                                        `p < ${calculatePValue([
                                            { value: parseInt(selectedPoll.resp_option1) },
                                            { value: parseInt(selectedPoll.resp_option2) },
                                        ]).toFixed(2)} (n=${parseInt(selectedPoll.resp_option1) + parseInt(selectedPoll.resp_option2)})`}
                                </span>
                            </div>
                        </CardContent>
                    </Card>

                    <Card id="data-table-card" className="flex flex-col overflow-hidden">
                        <CardHeader>
                            <CardTitle>Your Micro Surveys</CardTitle>
                        </CardHeader>
                        <CardContent id="data-table-content" className="flex flex-col flex-grow overflow-hidden">
                            <Input id="search-input" placeholder="Search..." value={searchTerm} onChange={handleSearch} className="mb-4" />
                            <div id="table-container" className="overflow-auto flex-grow">
                                <Table>
                                    <TableHeader>
                                        <TableRow>
                                            <TableHead>Question</TableHead>
                                            <TableHead>Option 1</TableHead>
                                            <TableHead>Option 2</TableHead>
                                        </TableRow>
                                    </TableHeader>
                                    <TableBody>
                                        {filteredData.map((poll) => (
                                            <TableRow
                                                key={poll.post_id}
                                                onClick={() => handleRowClick(poll)}
                                                className="cursor-pointer transition-colors hover:bg-gray-100 dark:hover:bg-gray-800"
                                            >
                                                <TableCell>{poll.caption}</TableCell>
                                                <TableCell>
                                                    {poll.option1} ({poll.resp_option1})
                                                </TableCell>
                                                <TableCell>
                                                    {poll.option2} ({poll.resp_option2})
                                                </TableCell>
                                            </TableRow>
                                        ))}
                                    </TableBody>
                                </Table>
                            </div>
                        </CardContent>
                    </Card>
                </div>
            </main>
            <footer id="dashboard-footer" className="flex-none p-4">
                <div id="button-container" className="flex justify-center space-x-4">
                    <Button id="function1-button" onClick={() => handleButtonClick("function1")}>
                        Function 1
                    </Button>
                    <Button id="function2-button" onClick={() => handleButtonClick("function2")}>
                        Function 2
                    </Button>
                    <Button id="function3-button" onClick={() => handleButtonClick("function3")}>
                        Function 3
                    </Button>
                    <Button id="function4-button" onClick={() => handleButtonClick("function4")}>
                        Function 4
                    </Button>
                    <Button id="function5-button" onClick={() => handleButtonClick("function5")}>
                        Function 5
                    </Button>
                </div>
            </footer>
        </div>
    );
}
