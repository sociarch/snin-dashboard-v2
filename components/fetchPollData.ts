import axios from "axios";
import { PollDataItem } from "./pollData";

const API_ENDPOINT = "https://0odsntgafl.execute-api.ap-southeast-1.amazonaws.com/Prod/codalookup";

export async function fetchPollData(): Promise<PollDataItem[]> {
    try {
        const response = await axios.get(API_ENDPOINT);

        if (response.data && response.data.success && Array.isArray(response.data.response)) {
            return response.data.response.map((item: any) => ({
                create_date: item.create_date || "",
                post_id: item.post_id || "",
                is_user_generated: item.is_user_generated || false,
                is_sponsored: item.is_sponsored || false,
                sponsor_id: item.sponsor_id || "",
                caption: item.caption || "Would you rather?",
                option1: item.option1 || "",
                option2: item.option2 || "",
                total_responses: item.total_responses || "",
                resp_option1: item.resp_option1 || "",
                pct_option1: item.pct_option1 || "",
                resp_option2: item.resp_option2 || "",
                pct_option2: item.pct_option2 || "",
                count_view: item.count_view || 0,
                count_like: item.count_like || 0,
                count_comment: item.count_comment || 0,
                count_bookmark: item.count_bookmark || 0,
                count_skip: item.count_skip || 0,
            }));
        } else {
            throw new Error("Invalid data structure received from the API");
        }
    } catch (error) {
        console.error("Error fetching poll data:", error);
        throw error;
    }
}
