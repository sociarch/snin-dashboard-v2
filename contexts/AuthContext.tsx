"use client";

import React, { createContext, useState, useContext, useEffect } from "react";
import { CognitoUser, AuthenticationDetails, CognitoUserAttribute } from "amazon-cognito-identity-js";
import { userPool } from "@/lib/awsConfig";

interface AuthContextType {
    isAuthenticated: boolean;
    user: CognitoUser | null;
    remainingQuestions: number | null;
    userAttributes: { [key: string]: string } | null;
    signIn: (username: string, password: string) => Promise<void>;
    signOut: () => void;
}

const AuthContext = createContext<AuthContextType | undefined>(undefined);

export const AuthProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
    const [isAuthenticated, setIsAuthenticated] = useState(false);
    const [user, setUser] = useState<CognitoUser | null>(null);
    const [remainingQuestions, setRemainingQuestions] = useState<number | null>(null);
    const [userAttributes, setUserAttributes] = useState<{ [key: string]: string } | null>(null);

    useEffect(() => {
        const currentUser = userPool.getCurrentUser();
        if (currentUser) {
            currentUser.getSession((err: any, session: any) => {
                if (err) {
                    console.error(err);
                } else if (session.isValid()) {
                    setIsAuthenticated(true);
                    setUser(currentUser);
                }
            });
        }
    }, []);

    const signIn = (username: string, password: string): Promise<void> => {
        return new Promise((resolve, reject) => {
            const authenticationDetails = new AuthenticationDetails({
                Username: username,
                Password: password,
            });

            const cognitoUser = new CognitoUser({
                Username: username,
                Pool: userPool,
            });

            cognitoUser.authenticateUser(authenticationDetails, {
                onSuccess: (result) => {
                    setIsAuthenticated(true);
                    setUser(cognitoUser);
                    console.log("Login response:", result);

                    cognitoUser.getUserAttributes((err, attributes) => {
                        if (err) {
                            console.error("Error getting user attributes:", err);
                            return;
                        }
                        if (attributes) {
                            const attrs: { [key: string]: string } = {};
                            attributes.forEach((attr) => {
                                attrs[attr.getName()] = attr.getValue();
                            });
                            setUserAttributes(attrs);
                            console.log("User attributes:", attrs);

                            const qsRemain = attributes.find((attr) => attr.getName() === "custom:qs_remain");
                            if (qsRemain) {
                                setRemainingQuestions(parseInt(qsRemain.getValue(), 10));
                            }
                        }
                    });

                    resolve();
                },
                onFailure: (err) => {
                    console.error("Login error:", err);
                    reject(err);
                },
            });
        });
    };

    const signOut = () => {
        if (user) {
            user.signOut();
            setIsAuthenticated(false);
            setUser(null);
            setUserAttributes(null);
        }
    };

    return <AuthContext.Provider value={{ isAuthenticated, user, remainingQuestions, userAttributes, signIn, signOut }}>{children}</AuthContext.Provider>;
};

export const useAuth = () => {
    const context = useContext(AuthContext);
    if (context === undefined) {
        throw new Error("useAuth must be used within an AuthProvider");
    }
    return context;
};
