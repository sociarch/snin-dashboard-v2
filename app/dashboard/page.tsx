"use client";

import { Dashboard } from "@/components/dashboard";
import { useAuth } from "@/contexts/AuthContext";
import { useRouter } from "next/navigation";
import { useEffect } from "react";

export default function DashboardPage() {
  const { isAuthenticated } = useAuth();
  const router = useRouter();

  useEffect(() => {
    if (!isAuthenticated) {
      router.push("/login");
    }
  }, [isAuthenticated, router]);

  return isAuthenticated ? <Dashboard /> : null;
}
