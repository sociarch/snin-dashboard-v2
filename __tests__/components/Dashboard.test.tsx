import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { Dashboard } from '@/components/dashboard';
import { useRouter } from 'next/navigation';
import '@testing-library/jest-dom';
import { AuthContext, AuthContextType } from '@/contexts/AuthContext';
import axios from 'axios';

// Mock axios
jest.mock('axios');
const mockedAxios = axios as jest.Mocked<typeof axios>;

// Mock next/navigation
jest.mock('next/navigation', () => ({
  useRouter: jest.fn()
}));

// Mock d3
jest.mock('d3', () => ({
  select: jest.fn(() => ({
    selectAll: jest.fn(() => ({
      remove: jest.fn()
    })),
    append: jest.fn(() => ({
      attr: jest.fn().mockReturnThis(),
      style: jest.fn().mockReturnThis(),
      append: jest.fn().mockReturnThis()
    }))
  })),
  arc: jest.fn(() => ({
    innerRadius: jest.fn().mockReturnThis(),
    outerRadius: jest.fn().mockReturnThis()
  })),
  pie: jest.fn(() => ({
    sort: jest.fn().mockReturnThis(),
    value: jest.fn().mockReturnThis(),
    startAngle: jest.fn().mockReturnThis(),
    endAngle: jest.fn().mockReturnThis()
  }))
}));

// Mock fetchPollData with a delay to simulate async behavior
jest.mock('@/components/fetchPollData', () => ({
  fetchPollData: jest.fn().mockImplementation(() => 
    Promise.resolve([{
      post_id: '1',
      caption: 'Test Question',
      option1: 'Option 1',
      option2: 'Option 2',
      resp_option1: '10',
      resp_option2: '20',
      pct_option1: '0.33',
      pct_option2: '0.67',
      sponsor_id: 'test_group'
    }])
  )
}));

// Mock BotpressEmbed component
jest.mock('@/components/BotpressEmbed', () => ({
  BotpressEmbed: () => null
}));

// Add this helper function at the top of the test file
const findTableCell = (text: string) => {
  return screen.getByRole('cell', { name: text });
};

const findChartTitle = (text: string) => {
  return screen.getByRole('heading', { 
    name: text,
    level: 2 
  });
};

describe('Dashboard Component', () => {
  const mockRouter = {
    push: jest.fn(),
    refresh: jest.fn()
  };

  const mockAuthContext: AuthContextType = {
    signOut: jest.fn().mockImplementation(() => Promise.resolve()),
    remainingQuestions: 5,
    userAttributes: {
      email: 'test@example.com',
      'custom:zipnum': '12345',
      'custom:qs_remain': '5',
      sub: 'test-sub'
    },
    userGroups: ['test_group'],
    isAuthenticated: true,
    user: null,
    signIn: jest.fn(),
    signUp: jest.fn(),
    confirmSignUp: jest.fn(),
    resendConfirmationCode: jest.fn(),
    forgotPassword: jest.fn(),
    forgotPasswordSubmit: jest.fn(),
    updateUserAttributes: jest.fn(),
    loading: false
  };

  beforeEach(() => {
    jest.clearAllMocks();
    (useRouter as jest.Mock).mockReturnValue(mockRouter);
    
    // Mock axios response for user groups
    mockedAxios.get.mockResolvedValue({
      data: {
        success: true,
        response: ['test_group']
      }
    });
  });

  const renderDashboard = () => {
    return render(
      <AuthContext.Provider value={mockAuthContext}>
        <Dashboard />
      </AuthContext.Provider>
    );
  };

  it('renders the dashboard title', () => {
    renderDashboard();
    expect(screen.getByText('Snap')).toBeInTheDocument();
    expect(screen.getByText('Input')).toBeInTheDocument();
  });

  it('loads and displays poll data', async () => {
    renderDashboard();
    await waitFor(() => {
      expect(screen.getByText('Test Question')).toBeInTheDocument();
    }, { timeout: 3000 });
  });

  it('handles search functionality', async () => {
    renderDashboard();
    
    await waitFor(() => {
      expect(screen.getByText('Test Question')).toBeInTheDocument();
    }, { timeout: 3000 });

    const searchInput = screen.getByPlaceholderText('Search...');
    fireEvent.change(searchInput, { target: { value: 'Test' } });
    
    await waitFor(() => {
      expect(screen.getByText('Test Question')).toBeInTheDocument();
    });
  });

  it('handles row click and shows chart', async () => {
    renderDashboard();
    
    await waitFor(() => {
      const row = findTableCell('Test Question');
      expect(row).toBeInTheDocument();
      fireEvent.click(row);

      const chartTitle = findChartTitle('Test Question');
      expect(chartTitle).toBeInTheDocument();

      expect(screen.getByText('33%')).toBeInTheDocument();
      expect(screen.getByText('67%')).toBeInTheDocument();
      expect(screen.getByText(/Most people prefer/)).toBeInTheDocument();
    }, { timeout: 3000 });
  });

  it('handles sign out', async () => {
    renderDashboard();
    const signOutButton = screen.getByTitle('Sign Out');
    fireEvent.click(signOutButton);
    await waitFor(() => {
      expect(mockAuthContext.signOut).toHaveBeenCalled();
      expect(mockRouter.push).toHaveBeenCalledWith('/login');
    });
  });
}); 