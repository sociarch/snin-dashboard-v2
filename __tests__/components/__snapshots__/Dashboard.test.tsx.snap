// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Component matches snapshot 1`] = `
<div>
  <div
    class="flex flex-col h-screen"
    id="dashboard-container"
  >
    <header
      class="flex-none p-4 flex justify-between items-center"
      id="dashboard-header"
    >
      <div
        class="flex items-center"
      >
        <h1
          class="text-xl font-semibold"
          id="dashboard-title"
        >
          <span
            class="text-[#ffd700]"
          >
            Snap
          </span>
          <span
            class="dark:text-white text-black"
          >
            Input
          </span>
        </h1>
      </div>
      <div
        class="flex items-center space-x-4"
        id="header-controls"
      >
        <button
          class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-10 w-10"
          id="sign-out-button"
          title="Sign Out"
        >
          <svg
            class="lucide lucide-log-out h-[1.2rem] w-[1.2rem]"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9 21H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h4"
            />
            <polyline
              points="16 17 21 12 16 7"
            />
            <line
              x1="21"
              x2="9"
              y1="12"
              y2="12"
            />
          </svg>
        </button>
      </div>
    </header>
    <main
      class="flex-grow overflow-hidden p-4 flex flex-col"
      id="dashboard-main"
    >
      <div
        class="grid grid-cols-1 lg:grid-cols-[minmax(auto,600px)_1fr] gap-4 flex-grow"
        id="dashboard-grid"
      >
        <div
          class="rounded-xl border shadow flex flex-col bg-black text-white w-full max-w-[600px] mx-auto"
          id="chart-card"
        >
          <div
            class="p-6 pt-0 flex-grow flex flex-col h-full transition-opacity duration-300 opacity-100"
            id="chart-content"
          >
            <div
              class="flex-1 flex items-center justify-center transition-opacity duration-500 opacity-0"
              id="chart-title-container"
            >
              <h2
                class="font-bold text-center px-4 text-3xl"
                id="chart-title"
              >
                Click a row to see detailed results
              </h2>
            </div>
            <div
              class="flex-1 flex justify-center items-center"
              id="chart-container"
            />
            <div
              class="flex justify-center items-center transition-opacity duration-500 opacity-0"
              id="chart-percentages"
            >
              <table
                class="w-[95%] mx-auto"
                id="percentages-table"
              >
                <tbody>
                  <tr
                    class="w-full"
                  >
                    <td
                      class="w-1/2 text-left font-['Roboto'] text-[2em] font-bold"
                      id="option1-percentage"
                      style="color: rgb(255, 215, 0); font-family: var(--font-roboto);"
                    />
                    <td
                      class="w-1/2 text-right font-['Roboto'] text-[2em] font-bold"
                      id="option2-percentage"
                      style="color: rgb(250, 128, 114); font-family: var(--font-roboto);"
                    />
                  </tr>
                </tbody>
              </table>
            </div>
            <div
              class="flex justify-center items-center py-4 transition-opacity duration-500 opacity-0"
              id="chart-annotation"
            >
              <span
                class="text-[1.5em] text-center"
                id="chart-footer-center"
              >
                Most people prefer
                 
              </span>
            </div>
            <div
              class="flex-1 flex justify-between items-end px-2"
              id="chart-footer"
            >
              <span
                class="text-xs"
                id="chart-footer-left"
              >
                snapinput.com
              </span>
              <span
                class="text-xs"
                id="chart-footer-right"
              />
            </div>
          </div>
        </div>
        <div
          class="rounded-xl border bg-card text-card-foreground shadow flex flex-col"
          id="data-table-card"
        >
          <div
            class="flex-col space-y-1.5 p-6 transition-opacity duration-300 opacity-0 hidden"
          >
            <h3
              class="font-semibold leading-none tracking-tight"
              id="data-table-title"
            >
              Your Micro Surveys
            </h3>
          </div>
          <div
            class="p-6 pt-0 flex-grow overflow-hidden relative"
            id="data-table-content"
          >
            <div
              class="flex items-center justify-center h-full"
              id="loading-spinner"
            >
              <div
                class="spinner"
              />
            </div>
            <div
              class="absolute inset-0 flex flex-col transition-opacity duration-300 opacity-0 pointer-events-none"
              id="report-container"
            >
              <div
                class="flex items-center justify-center h-full"
                id="generate-report-container"
              >
                <button
                  class="inline-flex items-center justify-center rounded-md ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2 text-2xl font-bold"
                  id="generate-report-button"
                >
                  Generate Report
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
      <footer
        class="mt-4"
        id="dashboard-footer"
      >
        <div
          class="flex justify-center space-x-2 max-w-3xl mx-auto"
          id="button-container"
        >
          <button
            class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 h-10 px-4 py-2 flex-1 bg-black hover:bg-gray-800 text-white border-4 border-gray-600 hover:border-[#ffd700] transition-colors duration-200 opacity-50"
            id="function1-button"
          >
            View Report
          </button>
          <button
            class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 h-10 px-4 py-2 flex-1 bg-black hover:bg-gray-800 text-white border-4 border-gray-600 hover:border-[#ffd700] transition-colors duration-200"
            id="function5-button"
          >
            Questions Left: 5
          </button>
        </div>
      </footer>
    </main>
    <div
      class="fixed bottom-4 left-4 flex items-center space-x-2"
    >
      <button
        aria-controls="radix-:r1a:"
        aria-expanded="false"
        aria-haspopup="dialog"
        class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 w-10"
        data-state="closed"
        type="button"
      >
        <svg
          class="lucide lucide-info h-[1.2rem] w-[1.2rem]"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="10"
          />
          <path
            d="M12 16v-4"
          />
          <path
            d="M12 8h.01"
          />
        </svg>
      </button>
      <button
        class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 w-10"
      >
        <svg
          class="lucide lucide-sun h-[1.2rem] w-[1.2rem]"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="4"
          />
          <path
            d="M12 2v2"
          />
          <path
            d="M12 20v2"
          />
          <path
            d="m4.93 4.93 1.41 1.41"
          />
          <path
            d="m17.66 17.66 1.41 1.41"
          />
          <path
            d="M2 12h2"
          />
          <path
            d="M20 12h2"
          />
          <path
            d="m6.34 17.66-1.41 1.41"
          />
          <path
            d="m19.07 4.93-1.41 1.41"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
